test/src/e2e.dynval/dv.double-vote.test.ts

```diff
@@ -0,0 +1,146 @@
+    const { height, view, step, blockHash, privKey, signerIdx } = data;
+    const encodableStep = (stepString => {
+        switch (stepString) {
+            case "propose":
+                return 0;
+            case "prevote":
+                return 1;
+            case "precommit":
+                return 2;
+            case "commit":
+                return 3;
+        }
+    })(step);
+    const encodableVoteStep = [
+        U64.ensure(height).toEncodeObject(),
+        U64.ensure(view).toEncodeObject(),
+        encodableStep
+    ];
+    const encodableBlockHash =
+        blockHash === null ? [] : [blockHash.toEncodeObject()];
+    const encodableVoteOn = [encodableVoteStep, encodableBlockHash];
+
+    const rlpVoteOn = RLP.encode(encodableVoteOn);
+    const messageForSchnorr = blake256(rlpVoteOn);
+    const schnorrSignature = signSchnorr(messageForSchnorr, privKey);
+    // pad because signSchnorr function does not guarantee the length of r and s to be 64.
+    const encodableSchnorrSignature = new H512(
+        schnorrSignature.r.padStart(64, "0") +
+            schnorrSignature.s.padStart(64, "0")
+    ).toEncodeObject();
+
+    return [
+        encodableVoteOn,
+        encodableSchnorrSignature,
+        U64.ensure(signerIdx).toEncodeObject()
+    ];
+}
+
+function createReportDoubleVoteTransaction(
+    sdk: SDK,
+    message1: RLP.Input,
+    message2: RLP.Input
+): Custom {
+    return sdk.core.createCustomTransaction({
+        handlerId: HANDLER_ID,
+        bytes: RLP.encode([REPORT_DOUBLE_VOTE_ACTION_ID, message1, message2])
+    });
+}
+
+const allDynValidators = originalValidators.slice(0, 4);
+const [alice, ...otherDynValidators] = allDynValidators;
+
+async function beforeBanCheck(sdk: SDK, bestBlockNumber: number) {
+    const validatorsBefore = (await stake.getPossibleAuthors(
+        sdk,
+        bestBlockNumber
+    ))!.map(platformAddr => platformAddr.toString());
+    expect(validatorsBefore.length).to.be.equals(allDynValidators.length);
+    expect(validatorsBefore).to.includes(alice.platformAddress.toString());
+    expect(validatorsBefore).contains.all.members(
+        allDynValidators.map(validator => validator.platformAddress.toString())
+    );
+}
+
+async function afterBanCheck(sdk: SDK, bestBlockNumber: number) {
+    const validatorsAfter = (await stake.getPossibleAuthors(
+        sdk,
+        bestBlockNumber
+    ))!.map(platformAddr => platformAddr.toString());
+
+    expect(validatorsAfter.length).to.be.equals(otherDynValidators.length);
+    expect(validatorsAfter).not.to.includes(alice.platformAddress.toString());
+    expect(validatorsAfter).to.contains.all.members(
+        otherDynValidators.map(validator =>
+            validator.platformAddress.toString()
+        )
+    );
+    const bannedAfter = (await stake.getBanned(sdk, bestBlockNumber)).map(
+        platformAddr => platformAddr.toString()
+    );
+    expect(bannedAfter).to.includes(alice.platformAddress.toString());
```

Comment:
There are some helper functions like `isBanned` in jail tests.
How about reusing them here?

Reply:
the original form of `isBanned` is not applicable directly, because we cannot designate `checkingNode` in it. I think we should clean up the similar functions later.

Reply:
I think we need some FIXME comments then. Would you add it here?

Reply:
Ok, I added it.